File: Daily-Streak/Day-03_Number-of-Ways-to-Paint-N-3-Grid/Solution.java
class Solution {
    public int numOfWays(int n) {
        final int MOD = 1000000007;

        // same -> ABA pattern count
        // diff -> ABC pattern count
        long same = 6;
        long diff = 6;

        // Process each row from 2 to n
        for (int row = 2; row <= n; row++) {

            // Transition to next row
            long newSame = (same * 3 + diff * 2) % MOD;
            long newDiff = (same * 2 + diff * 2) % MOD;

            // Update values
            same = newSame;
            diff = newDiff;
        }

        // Final answer
        return (int)((same + diff) % MOD);
    }
}

File: Daily-Streak/Day-03_Number-of-Ways-to-Paint-N-3-Grid/Solution.py
class Solution:
    def numOfWays(self, n: int) -> int:
        MOD = 10**9 + 7

        # same -> ABA pattern count
        # diff -> ABC pattern count
        same = 6
        diff = 6

        # Build DP row by row
        for _ in range(2, n + 1):

            # Apply transition rules
            new_same = (same * 3 + diff * 2) % MOD
            new_diff = (same * 2 + diff * 2) % MOD

            # Update for next iteration
            same = new_same
            diff = new_diff

        # Total valid colorings
        return (same + diff) % MOD
